packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,346 @@
+        visitType(typeArgument, area, true);
+      }
+
+      visitType(type.getNumberIndexType(), area, true);
+      visitType(type.getStringIndexType(), area, true);
+
+      type.getCallSignatures().forEach(signature => {
+        functionLikeType = true;
+        visitSignature(signature, area);
+      });
+
+      type.getConstructSignatures().forEach(signature => {
+        functionLikeType = true;
+        visitSignature(signature, area);
+      });
+    }
+
+    // Catch-all: operator types like `keyof T`
+    else if (isOperatorType(type)) {
+      visitType(type.type, area, false);
+    }
+
+    // Catch-all: generic type references like `Exclude<T, null>`
+    else if (type.aliasTypeArguments) {
+      visitTypesList(type.aliasTypeArguments, area, true);
+    }
+  }
+
+  function incrementIdentifierCount(
+    id: ts.Identifier,
+    area: IdentifierArea,
+    asRepeatedType: boolean,
+  ): void {
+    const value = asRepeatedType ? 2 : 1;
+    const identifierCount = identifierCounts.get(id) ?? {
+      general: 0,
+      return: 0,
+    };
+    identifierCount[area] += value;
+    identifierCounts.set(id, identifierCount);
+  }
+
+  function incrementTypeCount(type: ts.Type): number {
+    const count = (typeCounts.get(type) ?? 0) + 1;
+    typeCounts.set(type, count);
+    return count;
+  }
+
+  function visitSignature(
+    signature: ts.Signature | undefined,
+    area: IdentifierArea,
+  ): void {
+    if (!signature) {
+      return;
+    }
+
+    if (signature.thisParameter) {
+      visitType(checker.getTypeOfSymbol(signature.thisParameter), area, false);
+    }
+
+    for (const parameter of signature.parameters) {
+      visitType(checker.getTypeOfSymbol(parameter), area, false);
+    }
+
+    for (const typeParameter of signature.getTypeParameters() ?? []) {
+      visitType(typeParameter, area, false);
+    }
+
+    visitType(
+      checker.getTypePredicateOfSignature(signature)?.type ??
+        signature.getReturnType(),
+      'return',
+      false,
+    );
+  }
+
+  function visitSymbolsListOnce(
+    symbols: ts.Symbol[],
+    area: IdentifierArea,
+    asRepeatedType: boolean,
+  ): void {
+    if (visitedSymbolLists.has(symbols)) {
+      return;
+    }
+
+    visitedSymbolLists.add(symbols);
+
+    for (const symbol of symbols) {
+      visitType(checker.getTypeOfSymbol(symbol), area, asRepeatedType);
+    }
+  }
+
+  function visitTypesList(
+    types: readonly ts.Type[] | undefined,
+    area: IdentifierArea,
+    asRepeatedType: boolean,
+  ): void {
+    for (const type of types ?? []) {
+      visitType(type, area, asRepeatedType);
+    }
+  }
+}
+
+interface MappedType extends ts.ObjectType {
+  typeParameter?: ts.Type;
+}
+
+function isMappedType(type: ts.Type): type is MappedType {
+  return 'typeParameter' in type;
+}
+
+interface OperatorType extends ts.Type {
+  type: ts.Type;
+}
+
+function isOperatorType(type: ts.Type): type is OperatorType {
+  return 'type' in type && !!type.type;
+}
```

Comment:
Question in the open: is there some existing type predicate we should use? Or a new one to add to `ts-api-utils`? I couldn't find anything.